### DISCUSSION TITLE
Q: how to modularize Gluecodium bindings

### DISCUSSION THREAD

**Discussion by derolf** [heremaps/gluecodium#1526] [Q&A]
Assume you have two (or more) projects A and B, where B uses elements from A.

Is it possible to generate and build “separate” libraries/packages, where A can also be used without B, but also together with B?

For Dart, for example build a package A and a package B, where B has A as dependency.

I see that there is a global typerepository that might be clashing.

**Comment by Hsilgos** [comment]
It's possible to do this for Swift and Java out of the box, for Flutter it is not (yet?) implemented.
Examples for Swift and Java are in [cmake tests](https://github.com/heremaps/gluecodium/tree/master/cmake/tests/unit).
Locally I'm trying to solve this problem with script which additionally rewrites imports in generated code. Actually for Swift there is also post-generate step to add imports of dependencies.

**Comment by derolf** [comment]
Thanks, but which of the examples is showing for Java?

**Comment by Hsilgos** [comment]
> Thanks, but which of the examples is showing for Java?

If C++ is configured and generated correctly like in examples which I mentioned then In gradle you just specify which generated java sources belong to module A or B.
The main idea is that only one of the modules must contain common generated code (the code which responses for class caches, common type converters, etc.)

**Comment by derolf** [comment]
But does A still contain references to B (type repository)?

**Comment by Hsilgos** [comment]
> But does A still contain references to B (type repository)?

Yes. A is required for B to function normally.

**Comment by derolf** [comment]
I meant the other way around: Does A contain any references to B? (For example, the type repository for Dart is a single file that contains references to A and B)

**Comment by Hsilgos** [comment]
> I meant the other way around: Does A contain any references to B? (For example, the type repository for Dart is a single file that contains references to A and B)

If you mean language-level references, then no, A doesn't contain such references to B (in Java and Swift). A can work without B. B can't work without A because it has language level references to A.

If you mean memory-level references - then A contains runtime type repository which matches an object in Java/Swift and in C++, it also means that A can work without B, but not vice versa.

**Reply by derolf** [reply]
How can I generate the dart-ffi and java-jni cpp files into a folder per package? Right now they are all flat.

**Reply by Hsilgos** [reply]
If you make separate targets with separate gluecodium steps it should generate sources to separate folders. In our CMake tests are plenty examples

**Reply by derolf** [reply]
We made it work, but had to fix up some Dart stuff to create multiple type repositories, we merge at runtime.

**Reply by Hsilgos** [reply]
Locally I only changed imports after GLuecodium generation. Did you make anything else?
